icu4c/source/common/wintz.cpp

```diff
@@ -47,77 +48,60 @@ U_NAMESPACE_BEGIN
+    //  dynamicTZI.DaylightBias == 0)
 
-    LocalUResourceBundlePointer bundle(ures_openDirect(nullptr, "windowsZones", &status));
-    ures_getByKey(bundle.getAlias(), "mapTimezones", bundle.getAlias(), &status);
+    CharString winTZ;
+    UErrorCode status = U_ZERO_ERROR;
+    winTZ.appendInvariantChars(UnicodeString(TRUE, dynamicTZI.TimeZoneKeyName, -1), status)
+        .append('\0', status);
 
-    // convert from wchar_t* (UTF-16 on Windows) to char* (UTF-8).
-    u_strToUTF8(dynamicTZKeyName, UPRV_LENGTHOF(dynamicTZKeyName), nullptr,
-        reinterpret_cast<const UChar*>(dynamicTZI.TimeZoneKeyName), -1, &status);
+    // Map Windows Timezone name (non-localized) to ICU timezone ID (Olson timezone id).
+    LocalUResourceBundlePointer winTZBundle(ures_openDirect(nullptr, "windowsZones", &status));
+    ures_getByKey(winTZBundle.getAlias(), "mapTimezones", winTZBundle.getAlias(), &status);
+    ures_getByKey(winTZBundle.getAlias(), winTZ.data(), winTZBundle.getAlias(), &status);
 
-    if (U_FAILURE(status)) {
+    if (U_FAILURE(status))
         return nullptr;
-    }
 
-    if (dynamicTZI.TimeZoneKeyName[0] != 0) {
-        StackUResourceBundle winTZ;
-        ures_getByKey(bundle.getAlias(), dynamicTZKeyName, winTZ.getAlias(), &status);
-
-        if (U_SUCCESS(status)) {
-            const UChar* icuTZ = nullptr;
-            if (errorCode != 0) {
-                icuTZ = ures_getStringByKey(winTZ.getAlias(), ISOcode, &len, &status);
-            }
-            if (errorCode == 0 || icuTZ == nullptr) {
-                /* fallback to default "001" and reset status */
-                status = U_ZERO_ERROR;
-                icuTZ = ures_getStringByKey(winTZ.getAlias(), "001", &len, &status);
-            }
-
-            if (U_SUCCESS(status)) {
-                int index = 0;
-
-                while (!(*icuTZ == '\0' || *icuTZ == ' ')) {
-                    // time zone IDs only contain ASCII invariant characters.
-                    tmpid[index++] = (char)(*icuTZ++);
-                }
-                tmpid[index] = '\0';
-            }
-        }
+    const UChar* icuTZ16 = nullptr;
+    char regionCode[3] = {}; // 2 letter ISO 3166 country code
+    int geoId = GetUserGeoID(GEOCLASS_NATION);
+    int regionCodeLen = GetGeoInfoA(geoId, GEO_ISO2, regionCode, 3, 0);
+    int32_t tzLen;
+    if (regionCodeLen != 0) {
+        icuTZ16 = ures_getStringByKey(winTZBundle.getAlias(), regionCode, &tzLen, &status);
     }
-
-    // Copy the timezone ID to icuid to be returned.
-    if (tmpid[0] != 0) {
-        icuid = uprv_strdup(tmpid);
+    if (regionCodeLen == 0 || U_FAILURE(status)) {
+        // fallback to default "001" (world)
+        status = U_ZERO_ERROR;
+        icuTZ16 = ures_getStringByKey(winTZBundle.getAlias(), "001", &tzLen, &status);
     }
 
-    return icuid;
+    CharString icuTZStr;
+    return icuTZStr.appendInvariantChars(icuTZ16, tzLen, status).cloneData(status);
```

Comment:
Ick..... I wish appendInvariantChars did set |status| to indicate OOM, but it does not.  And, cloneData() returns "" instead of nullptr for an empty string.  

BTW, none of IANA timezone id is longer than the size of CharString's internal stack buffer size (40; none of them is even 36 chars long). So, it'll never OOM.   

Nonethless, in the future, an ID longer than 40 chars can be added. So, I'll add a check. I'll also file a "bug" about CharString to indicate OOM in append*().

Reply:
The same here. I was mistaken about EnsureCapacity inside appendInvariantChars while responding to your comment.  The original PR should work because cloneData() returns nullptr for U_FAILURE(status).

Reply:
Sorry for not catching/realizing that `cloneData()` returns `nullptr`. Yes, looking back at your original change, I think it would work. :)